### ISSUE TITLE
Long-polling or other long running handlers delay shutdown

### ISSUE THREAD

**Issue by miguelgrinberg** [Kludex/uvicorn#251]
Here is an little example app using Quart:

```python
import asyncio
from quart import Quart

app = Quart(__name__)

@app.route('/')
async def index():
    await asyncio.sleep(60)
    return 'foo'
```

Start the app with uvicorn, open it in the browser and then go back to the terminal and hit Ctrl-C on the uvicorn process. The process will take about a minute to exit. Hitting Ctrl-C again does not help.

As a reference point, Daphne aborts the running handler and promptly exists back to the prompt. Hypercorn hangs like uvicorn, but hitting Ctrl-C a second time releases the process immediately.

**Comment by lovelydinosaur** [comment]
Okay, so this is actually a little involved.

Shutdown needs to be graceful, and should avoid hard-dropping connections wherever possible. After all, typically shutdown isn't "the server is going away", but rather "this particular process is being restarted". In a robust setup you'd expect individual processes to perform restarts perhaps every 24hrs or so, to help  guard against resource leaks.

I'll walk through what the existing shutdown behavior is first...

* Stop accepting new connections.
* Close any connections that are not currently in a request/response cycle, or running a background task. (ie. a response has been sent, but the ASGI task has not yet completed)
* Wait for any connections in a request/response cycle to terminate.
* Wait for any background tasks to terminate.
* Fire the shutdown event and wait for it to terminate.

There are two changes that we ought to make to this behavior.

* A second Ctrl-C should just hard-exit.
* Shutdown ought to enqueue an ASGI `disconnect` message so that long-polling connections are able to react to the event. The ASGI spec will probably need to be precise here to make it clear that a disconnect does not necessarily mean that the socket has been hard-closed, just that it is closing. Standard request-responses won't bother to monitor for the message, and don't need to pay attention to it in any case, they should just send their response to completion and return normally - so long as they do so within a reasonable timeout.

Something to be noted here, is that even after these two changes, the example you've given above wouldn't get graceful shutdown (but would exit on a second Ctrl-C). For long-polling (and indeed for websockets) we always ought to be structuring our endpoints in an event-driven manner, rather than hanging a single flow of control of a main entry point.

There's more to be said there, about how broadcast messaging ought to work for websockets and long-polling, that simply isn't properly covered yet by either Quart or Starlette. (Happy to dive into that with you, but it'd probably help if I first get some more context)

**Comment by miguelgrinberg** [comment]
I apologize for the terrible example I provided. This was just an attempt to simplify the problem that I'm seeing, but I went too far. My example is a SocketIO long-polling server, and the polling cycle has a timeout of 60 seconds by default. Clients send a ping message every 25 seconds approximately and the server responds with a pong immediately, breaking the polling cycle. So during normal use a polling request is at the most 25 or so seconds long. But obviously when I hit Ctrl-C the pings stop being delivered, so the last polling request stays on for the whole 60 seconds.

A second Ctrl-C doing a hard exit makes sense to me. Hypercorn is doing that, and I don't remember which of gevent or eventlet also does that with my server.

Getting a `disconnect` or `shutdown` event would definitely help, as I can then halt the polling request immediately and clean up after myself.
